internal/ceb/ssh/server.go

```diff
@@ -0,0 +1,253 @@
+	ctx context.Context,
+	logger hclog.Logger,
+	sport string,
+	hostkey gossh.Signer,
+	key gossh.PublicKey,
+) error {
+	var server *ssh.Server
+
+	port, err := strconv.Atoi(sport)
+	if err != nil {
+		return err
+	}
+
+	check := func(ctx ssh.Context, inputKey ssh.PublicKey) bool {
+		if ssh.KeysEqual(inputKey, key) {
+			return true
+		}
+
+		logger.Error("keys did not match")
+		return false
+	}
+
+	logger.Info("starting ssh listener...")
+
+	err = ssh.ListenAndServe(
+		fmt.Sprintf(":%d", port),
+		createHandler(ctx, logger, &server),
+		ssh.Option(func(serv *ssh.Server) error {
+			server = serv
+			serv.PublicKeyHandler = check
+			serv.AddHostKey(hostkey)
+			return nil
+		}),
+	)
+
+	if err != ssh.ErrServerClosed {
+		return err
+	}
+
+	return nil
+}
+
+// This is pulled out to make it easier to test.
+func createHandler(ctx context.Context, logger hclog.Logger, server **ssh.Server) ssh.Handler {
+	return func(s ssh.Session) {
+		// Build the subprocess
+
+		args := s.Command()
+
+		cmd := exec.CommandContext(s.Context(), args[0], args[1:]...)
+		cmd.Env = s.Environ()
+
+		var (
+			ptyFile *os.File
+			err     error
+		)
+
+		logger.Debug("executing command", "command", args)
+
+		if ptyInfo, winCh, ok := s.Pty(); ok {
+			logger.Debug("running command in a PTY")
+
+			// If we're setting a pty we'll be overriding our stdin/out/err
+			// so we need to get access to the original gRPC writers so we can
+			// copy later.
+			stdin := s
+			stdout := s
+
+			// Set our TERM value
+			if ptyInfo.Term != "" {
+				cmd.Env = append(cmd.Env, "TERM="+ptyInfo.Term)
+			}
+
+			// pty.StartWithSize sets "setsid" which is mutually exclusive to
+			// Setpgid. They both result in a new process group being created with
+			// the process group ID equal to the PID, which is the behavior we
+			// expect when terminating processes.
+			if cmd.SysProcAttr != nil {
+				cmd.SysProcAttr.Setpgid = false
+			}
+
+			// Start with a pty
+			ptyFile, err = pty.StartWithSize(cmd, &pty.Winsize{
+				X: uint16(ptyInfo.Window.Width),
+				Y: uint16(ptyInfo.Window.Height),
+			})
+			if err != nil {
+				fmt.Fprintf(s, "Error occured: %s\r\n", err)
+				return
+			}
+
+			defer ptyFile.Close()
+
+			// Copy stdin to the pty
+			go func() {
+				io.Copy(ptyFile, stdin)
+				logger.Debug("ssh client closed stdin")
+			}()
+
+			go func() {
+				io.Copy(stdout, ptyFile)
+				logger.Debug("command closed stdout")
+			}()
+
+			go func() {
+				for {
+					select {
+					case <-s.Context().Done():
+						return
+
+					case win := <-winCh:
+						sz := pty.Winsize{
+							X: uint16(win.Width),
+							Y: uint16(win.Height),
+						}
+
+						if err := pty.Setsize(ptyFile, &sz); err != nil {
+							logger.Warn("error changing window size, this doesn't quit the stream",
+								"err", err)
+						}
+					}
+				}
+			}()
+		} else {
+			logger.Debug("executing command without pty")
+
+			// You might think "Hey, why make these pipes, can't we just
```

Comment:
Appreciate these kinds of comments :smile: :+1:

Reply:
Luckily I had hit this many years ago and when I hit it again, it didn't take long to remember. But it's EXTREMELY unclear what is happening.